// xdl/build/Credentials.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getExistingDistCerts = exports.removeCredentialsForPlatform = exports.updateCredentialsForPlatform = exports.getCredentialsForPlatform = exports.credentialsExistForPlatformAsync = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

let credentialsExistForPlatformAsync = exports.credentialsExistForPlatformAsync = (() => {
  var _ref = _asyncToGenerator(function* (metadata) {
    return fetchCredentials(metadata, false);
  });

  return function credentialsExistForPlatformAsync(_x) {
    return _ref.apply(this, arguments);
  };
})();

let getCredentialsForPlatform = exports.getCredentialsForPlatform = (() => {
  var _ref2 = _asyncToGenerator(function* (metadata) {
    return fetchCredentials(metadata, true);
  });

  return function getCredentialsForPlatform(_x2) {
    return _ref2.apply(this, arguments);
  };
})();

let fetchCredentials = (() => {
  var _ref3 = _asyncToGenerator(function* ({ username, experienceName, bundleIdentifier, platform }, decrypt) {
    // this doesn't hit our mac rpc channel, so it needs significantly less debugging
    const { err, credentials } = yield (_Api || _load_Api()).default.callMethodAsync('getCredentials', [], 'post', {
      username,
      experienceName,
      bundleIdentifier,
      platform,
      decrypt
    });

    if (err) {
      throw new Error('Error fetching credentials.');
    }

    return credentials;
  });

  return function fetchCredentials(_x3, _x4) {
    return _ref3.apply(this, arguments);
  };
})();

let updateCredentialsForPlatform = exports.updateCredentialsForPlatform = (() => {
  var _ref4 = _asyncToGenerator(function* (platform, newCredentials, metadata) {
    // this doesn't go through the mac rpc, no request id needed
    const { err, credentials } = yield (_Api || _load_Api()).default.callMethodAsync('updateCredentials', [], 'post', _extends({
      credentials: newCredentials,
      platform
    }, metadata));

    if (err || !credentials) {
      throw new Error('Error updating credentials.');
    }
  });

  return function updateCredentialsForPlatform(_x5, _x6, _x7) {
    return _ref4.apply(this, arguments);
  };
})();

let removeCredentialsForPlatform = exports.removeCredentialsForPlatform = (() => {
  var _ref5 = _asyncToGenerator(function* (platform, metadata) {
    // doesn't go through mac rpc, no request id needed
    const { err } = yield (_Api || _load_Api()).default.callMethodAsync('deleteCredentials', [], 'post', _extends({
      platform
    }, metadata));

    if (err) {
      throw new Error('Error deleting credentials.');
    }
  });

  return function removeCredentialsForPlatform(_x8, _x9) {
    return _ref5.apply(this, arguments);
  };
})();

let getExistingDistCerts = exports.getExistingDistCerts = (() => {
  var _ref6 = _asyncToGenerator(function* (username, appleTeamId) {
    const { err, certs } = yield (_Api || _load_Api()).default.callMethodAsync('getExistingDistCerts', [], 'post', {
      username,
      appleTeamId
    });

    if (err) {
      throw new Error('Error getting existing distribution certificates.');
    }

    return certs.map(function (_ref7) {
      let { usedByApps } = _ref7,
          rest = _objectWithoutProperties(_ref7, ['usedByApps']);

      return _extends({
        usedByApps: usedByApps && usedByApps.split(';')
      }, rest);
    });
  });

  return function getExistingDistCerts(_x10, _x11) {
    return _ref6.apply(this, arguments);
  };
})();

var _Api;

function _load_Api() {
  return _Api = _interopRequireDefault(require('./Api'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }
