// exp/build/commands/fetch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var writeKeystore = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(projectDir) {
    var temporary = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

    var _ref2, _ref2$args, username, remotePackageName, experienceName, outputFile, credentialMetadata, credentials, keystore, keystorePassword, keyAlias, keyPassword, storeBuf;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (_xdl || _load_xdl()).Exp.getPublishInfoAsync(projectDir);

          case 2:
            _ref2 = _context.sent;
            _ref2$args = _ref2.args;
            username = _ref2$args.username;
            remotePackageName = _ref2$args.remotePackageName;
            experienceName = _ref2$args.remoteFullPackageName;
            outputFile = _path.default.resolve(projectDir, '' + remotePackageName + (temporary ? '.tmp' : '') + '.jks');
            credentialMetadata = { username: username, experienceName: experienceName, platform: 'android' };


            (0, (_log || _load_log()).default)('Retreiving Android keystore for ' + experienceName);
            _context.next = 12;
            return (_xdl || _load_xdl()).Credentials.getCredentialsForPlatform(credentialMetadata);

          case 12:
            credentials = _context.sent;

            if (credentials) {
              _context.next = 15;
              break;
            }

            throw new Error('Unable to fetch credentials for this project. Are you sure they exist? Did you build a standalone Android app?');

          case 15:
            keystore = credentials.keystore, keystorePassword = credentials.keystorePassword, keyAlias = credentials.keystoreAlias, keyPassword = credentials.keyPassword;
            storeBuf = Buffer.from(keystore, 'base64');


            if (!temporary) {
              (0, (_log || _load_log()).default)('Writing keystore to ' + outputFile + '...');
            }
            _fs.default.writeFileSync(outputFile, storeBuf);

            return _context.abrupt('return', {
              outputFile: outputFile,
              keystorePassword: keystorePassword,
              keyPassword: keyPassword,
              keyAlias: keyAlias
            });

          case 20:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function writeKeystore(_x) {
    return _ref.apply(this, arguments);
  };
}();

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _fs = _interopRequireDefault(require('fs'));

var _path = _interopRequireDefault(require('path'));

var _xdl;

function _load_xdl() {
  return _xdl = require('xdl');
}

var _crypto = _interopRequireDefault(require('crypto'));

var _spawnAsync;

function _load_spawnAsync() {
  return _spawnAsync = _interopRequireDefault(require('@expo/spawn-async'));
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../log'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function exportCertAsync(keystoreFile, keystorePassword, keyAlias, certFile) {
  return (0, (_spawnAsync || _load_spawnAsync()).default)('keytool', ['-exportcert', '-keystore', keystoreFile, '-storepass', keystorePassword, '-alias', keyAlias, '-file', certFile, '-noprompt', '-storetype', 'JKS']);
}

exports.default = function (program) {
  program.command('fetch:ios:certs [project-dir]').description('Fetch this project\'s iOS certificates and provisioning profile. Writes certificates to PROJECT_DIR/PROJECT_NAME_(dist|push).p12 and prints passwords to stdout.').asyncActionProjectDir(function () {
    var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(projectDir, options) {
      var _ref4, _ref4$args, username, remotePackageName, experienceName, bundleIdentifier, distOutputFile, pushOutputFile, credentialMetadata, _ref5, certP12, certPassword, certPrivateSigningKey, pushP12, pushPassword, pushPrivateSigningKey, provisioningProfile, teamId, keyPath, _keyPath, p;

      return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return (_xdl || _load_xdl()).Exp.getPublishInfoAsync(projectDir);

            case 2:
              _ref4 = _context2.sent;
              _ref4$args = _ref4.args;
              username = _ref4$args.username;
              remotePackageName = _ref4$args.remotePackageName;
              experienceName = _ref4$args.remoteFullPackageName;
              bundleIdentifier = _ref4$args.bundleIdentifierIOS;
              distOutputFile = _path.default.resolve(projectDir, remotePackageName + '_dist.p12');
              pushOutputFile = _path.default.resolve(projectDir, remotePackageName + '_push.p12');
              credentialMetadata = { username: username, experienceName: experienceName, platform: 'ios', bundleIdentifier: bundleIdentifier };


              (0, (_log || _load_log()).default)('Retreiving iOS credentials for ' + experienceName);

              _context2.prev = 12;
              _context2.next = 15;
              return (_xdl || _load_xdl()).Credentials.getCredentialsForPlatform(credentialMetadata);

            case 15:
              _ref5 = _context2.sent;
              certP12 = _ref5.certP12;
              certPassword = _ref5.certPassword;
              certPrivateSigningKey = _ref5.certPrivateSigningKey;
              pushP12 = _ref5.pushP12;
              pushPassword = _ref5.pushPassword;
              pushPrivateSigningKey = _ref5.pushPrivateSigningKey;
              provisioningProfile = _ref5.provisioningProfile;
              teamId = _ref5.teamId;

              // if undefines because some people might have pre-local-auth as default credentials.
              if (teamId !== undefined) {
                (0, (_log || _load_log()).default)('These credentials are associated with Apple Team ID: ' + teamId);
              }
              (0, (_log || _load_log()).default)('Writing distribution cert to ' + distOutputFile + '...');
              _fs.default.writeFileSync(distOutputFile, Buffer.from(certP12, 'base64'));
              if (certPrivateSigningKey !== undefined) {
                keyPath = _path.default.resolve(projectDir, remotePackageName + '_dist_cert_private.key');

                _fs.default.writeFileSync(keyPath, certPrivateSigningKey);
              }
              (0, (_log || _load_log()).default)('Done writing distribution cert credentials to disk.');
              (0, (_log || _load_log()).default)('Writing push cert to ' + pushOutputFile + '...');
              _fs.default.writeFileSync(pushOutputFile, Buffer.from(pushP12, 'base64'));
              if (pushPrivateSigningKey !== undefined) {
                _keyPath = _path.default.resolve(projectDir, remotePackageName + '_push_cert_private.key');

                _fs.default.writeFileSync(_keyPath, pushPrivateSigningKey);
              }
              (0, (_log || _load_log()).default)('Done writing push cert credentials to disk.');
              if (provisioningProfile !== undefined) {
                p = _path.default.resolve(projectDir, remotePackageName + '.mobileprovision');

                (0, (_log || _load_log()).default)('Writing provisioning profile to ' + p + '...');
                _fs.default.writeFileSync(p, Buffer.from(provisioningProfile, 'base64'));
                (0, (_log || _load_log()).default)('Done writing provisioning profile to disk');
              }
              (0, (_log || _load_log()).default)('Save these important values as well:\n\nDistribution p12 password: ' + (_chalk || _load_chalk()).default.bold(certPassword) + '\nPush p12 password:         ' + (_chalk || _load_chalk()).default.bold(pushPassword) + '\n');
              _context2.next = 40;
              break;

            case 37:
              _context2.prev = 37;
              _context2.t0 = _context2['catch'](12);
              throw new Error('Unable to fetch credentials for this project. Are you sure they exist?');

            case 40:

              (0, (_log || _load_log()).default)('All done!');

            case 41:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, undefined, [[12, 37]]);
    }));

    return function (_x3, _x4) {
      return _ref3.apply(this, arguments);
    };
  }(), true);

  program.command('fetch:android:keystore [project-dir]').description("Fetch this project's Android keystore. Writes keystore to PROJECT_DIR/PROJECT_NAME.jks and prints passwords to stdout.").asyncActionProjectDir(function () {
    var _ref6 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee3(projectDir, options) {
      var _ref7, keystorePassword, keyPassword, keyAlias;

      return (_regenerator || _load_regenerator()).default.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return writeKeystore(projectDir);

            case 2:
              _ref7 = _context3.sent;
              keystorePassword = _ref7.keystorePassword;
              keyPassword = _ref7.keyPassword;
              keyAlias = _ref7.keyAlias;


              (0, (_log || _load_log()).default)('Done writing keystore to disk.');
              (0, (_log || _load_log()).default)('Save these important values as well:\n\nKeystore password: ' + (_chalk || _load_chalk()).default.bold(keystorePassword) + '\nKey alias:         ' + (_chalk || _load_chalk()).default.bold(keyAlias) + '\nKey password:      ' + (_chalk || _load_chalk()).default.bold(keyPassword) + '\n');

              (0, (_log || _load_log()).default)('All done!');

            case 9:
            case 'end':
              return _context3.stop();
          }
        }
      }, _callee3, undefined);
    }));

    return function (_x5, _x6) {
      return _ref6.apply(this, arguments);
    };
  }(), true);

  program.command('fetch:android:hashes [project-dir]').description("Fetch this project's Android key hashes needed to setup Google/Facebook authentication.").asyncActionProjectDir(function () {
    var _ref8 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee4(projectDir, options) {
      var _ref9, outputFile, keystorePassword, keyAlias, certFile, data, googleHash, fbHash;

      return (_regenerator || _load_regenerator()).default.wrap(function _callee4$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return writeKeystore(projectDir, true);

            case 2:
              _ref9 = _context4.sent;
              outputFile = _ref9.outputFile;
              keystorePassword = _ref9.keystorePassword;
              keyAlias = _ref9.keyAlias;
              certFile = outputFile.replace('jks', 'cer');
              _context4.prev = 7;
              _context4.next = 10;
              return exportCertAsync(outputFile, keystorePassword, keyAlias, certFile);

            case 10:
              data = _fs.default.readFileSync(certFile);
              googleHash = _crypto.default.createHash('sha1').update(data).digest('hex').toUpperCase();
              fbHash = _crypto.default.createHash('sha1').update(data).digest('base64');

              (0, (_log || _load_log()).default)('Google Certificate Fingerprint:  ' + googleHash.replace(/(.{2}(?!$))/g, '$1:'));
              (0, (_log || _load_log()).default)('Google Certificate Hash:         ' + googleHash);
              (0, (_log || _load_log()).default)('Facebook Key Hash:               ' + fbHash);
              _context4.next = 24;
              break;

            case 18:
              _context4.prev = 18;
              _context4.t0 = _context4['catch'](7);

              if (_context4.t0.code === 'ENOENT') {
                (_log || _load_log()).default.warn('Are you sure you have keytool installed?');
                (0, (_log || _load_log()).default)('keytool is part of openJDK: http://openjdk.java.net/');
                (0, (_log || _load_log()).default)('Also make sure that keytool is in your PATH after installation.');
              }
              if (_context4.t0.stdout) {
                (0, (_log || _load_log()).default)(_context4.t0.stdout);
              }
              if (_context4.t0.stderr) {
                (_log || _load_log()).default.error(_context4.t0.stderr);
              }
              throw _context4.t0;

            case 24:
              _context4.prev = 24;

              try {
                _fs.default.unlinkSync(certFile);
              } catch (err) {
                if (err.code !== 'ENOENT') {
                  (_log || _load_log()).default.error(err);
                }
              }
              try {
                _fs.default.unlinkSync(outputFile);
              } catch (err) {
                if (err.code !== 'ENOENT') {
                  (_log || _load_log()).default.error(err);
                }
              }
              return _context4.finish(24);

            case 28:

              (0, (_log || _load_log()).default)('All done!');

            case 29:
            case 'end':
              return _context4.stop();
          }
        }
      }, _callee4, undefined, [[7, 18, 24, 28]]);
    }));

    return function (_x7, _x8) {
      return _ref8.apply(this, arguments);
    };
  }(), true);
};

module.exports = exports['default'];
